File: game/readied.py
"""Held actions expire with the round; failed Dexterity shifts the next round."""
from types import SimpleNamespace
import uuid
from .rules import availability, computed, current_scene, Change


def held_actions(c):
    return ([c.runtime['readied']] if c.runtime.get('readied') else [])+c.runtime.get('readied_queue',[])


def selected(c,action=None,key=None):
    return next((h for h in held_actions(c) if (key is None or h.get('key')==key) and (action is None or h['action']==action)),None)


def reason(c,scene,action,key=None):
    held=selected(c,action,key)
    if not scene or not held or held['round']!=scene.state['round']:
        return 'Нет подходящего отложенного действия'
    return ''


def reserve(user,p):
    from .views import owned
    c=owned(user,p['character']);scene=current_scene(c);key=p.get('action')
    if key not in ['main','move','minor']:raise ValueError('Выберите основное, движение или малое действие')
    blocked=availability(c,SimpleNamespace(data={'action':key}),scene)
    if blocked:raise ValueError(blocked)
    if c.runtime.get('actions',{}).get('minor',0)<(2 if key=='minor' else 1):
        raise ValueError('Чтобы отложить действие, нужно потратить ещё одно малое')
    condition=str(p.get('condition','')).strip()
    if not condition or len(condition)>500:raise ValueError('Опишите условие срабатывания — до 500 символов')
    change=Change(user,'Отложенное действие · '+c.name,scene,inputs={'condition':condition,'action':key})
    change.watch(c)
    c.runtime['actions']['minor']-=1;c.runtime['actions'][key]-=1
    held={'key':str(uuid.uuid4()),'action':key,'condition':condition,'round':scene.state['round']}
    if c.runtime.get('readied'):c.runtime.setdefault('readied_queue',[]).append(held)
    else:c.runtime['readied']=held
    change.finish()


def resolve(change,c,scene,p,action=None):
    held=selected(c,action,p.get('ready_id'))
    if not held or held['round']!=scene.state['round']:raise ValueError('Отложенное действие истекло')
    if p.get('triggered') is not True:raise ValueError('Подтвердите наступление условия')
    fail=p.get('voluntary_fail') is True
    roll=p.get('dex_roll')
    if not fail and (type(roll) is not int or not -1000<=roll<=1000):raise ValueError('Введите физический результат проверки Ловкости или выберите добровольный провал')
    bonus=computed(c)['mods']['dex'];total=None if fail else roll+bonus
    trigger=scene.state['order'][scene.state['turn']]
    change.watch(c)
    if fail or total<10:
        c.runtime['initiative_shift']={'after':trigger,'value':scene.state.get('initiative',{}).get(str(trigger),0)-1}
    check='Ловкость: добровольный провал' if fail else f'Ловкость: {roll} {bonus:+d} = {total} / 10'
    change.inputs['roll_result']=(change.inputs.get('roll_result','')+' · '+check).strip(' ·')
    change.inputs['readied']={**held,'dex_roll':None if fail else roll,'bonus':bonus,'total':total,'voluntary_fail':fail,'trigger':trigger}
    remaining=[h for h in held_actions(c) if h is not held]
    c.runtime.pop('readied',None);c.runtime.pop('readied_queue',None)
    if remaining:c.runtime['readied']=remaining[0];c.runtime['readied_queue']=remaining[1:]


def perform(user,p):
    from .views import owned
    c=owned(user,p['character']);scene=current_scene(c)
    if not scene:raise ValueError('Персонаж вне боя')
    held=selected(c,key=p.get('ready_id'))
    blocked=availability(c,SimpleNamespace(data={'action':held['action'] if held else None}),scene,readied=p.get('ready_id') or True)
    if blocked:raise ValueError(blocked)
    change=Change(user,'Выполнено отложенное действие · '+c.name,scene)
    resolve(change,c,scene,p)
    change.finish()


def next_round(scene,chars):
    order=scene.state['order'];shifts=[]
    for c in chars.values():
        c.runtime.pop('readied',None);c.runtime.pop('readied_queue',None)
        shift=c.runtime.pop('initiative_shift',None)
        if shift:
            scene.state.setdefault('initiative',{})[str(c.id)]=shift['value']
            shifts.append((c.id,shift['after']))
    for pk,after in shifts:
        if pk!=after and pk in order and after in order:
            order.remove(pk);order.insert(order.index(after)+1,pk)

File: static/readied.js
function heldActions(c){return [...(c.runtime.readied?[c.runtime.readied]:[]),...(c.runtime.readied_queue||[])]}
function readyFields(c,key=null){const held=heldActions(c).find(h=>!key||h.key===key);return `<section><p><strong>Условие:</strong> ${esc(held?.condition||'')}</p><label class="checks"><input type="checkbox" name="triggered" required>Условие наступило</label><h3>Проверка после действия · Ловкость 10</h3><p>Модификатор Ловкости: ${c.calc.mods.dex>=0?'+':''}${c.calc.mods.dex}</p>${input('dex_roll','Физический результат без модификатора','','number','min="-1000" max="1000" required')}<label class="checks"><input type="checkbox" name="voluntary_fail">Добровольный провал</label></section>`}
function readyPayload(a,f){return a?.use_ready?{use_ready:true,ready_id:a.ready_id,triggered:!!f.triggered,voluntary_fail:!!f.voluntary_fail,dex_roll:f.dex_roll===''?null:Number(f.dex_roll)}:{}}
function readyPanel(c,scene){if(!scene)return '';return heldActions(c).map(held=>`<details class="card" open><summary>Отложено: ${esc(actionNames[held.action])}</summary><p>${esc(held.condition)}</p><small>До конца раунда ${held.round}</small>${c.editable?`<div class="button-group">${btn('Применить умение','ready.ability',`data-ready="${esc(held.key)}"`)}${btn('Выполнить действие','ready.perform',`data-ready="${esc(held.key)}"`)}</div>`:''}</details>`).join('')+(c.editable?btn('Отложить действие','ready.reserve',` ${skippedConditions(c).length||scene.state.order[scene.state.turn]!==c.id||!c.runtime.actions?.minor?'disabled':''}`):'')}

function reserveForm(c){
 const actions=['main','move','minor'].filter(k=>(c.runtime.actions?.[k]||0)>(k==='minor'?1:0));
 modal('Отложить действие',`<p>Расходуется малое действие; выбранное действие резервируется до конца раунда.</p>${selector('ready_action','Действие',actions.map(id=>({id,name:actionNames[id]})),actions[0])}${input('ready_condition','Условие срабатывания','','text','required maxlength="500"')}`,async()=>api({op:'action.ready',character:c.id,action:formObject().ready_action,condition:formObject().ready_condition}),'Отложить');
 el('dialog-submit').disabled=!actions.length;
}
document.addEventListener('click',e=>{
 const b=e.target.closest('[data-do]');if(!b)return;const c=current();
 if(b.dataset.do==='ready.reserve')reserveForm(c);
 if(b.dataset.do==='ready.perform')modal('Выполнить отложенное действие',readyFields(c,b.dataset.ready),async()=>api({op:'action.perform_ready',character:c.id,...readyPayload({use_ready:true,ready_id:b.dataset.ready},formObject())}),'Выполнить');
 if(b.dataset.do==='ready.ability'){
  const held=heldActions(c).find(h=>h.key===b.dataset.ready),choices=c.abilities.filter(a=>!a.ready_reason&&(a.data.category||'active')==='active'&&(a.data.action||'main')===held?.action);
  modal('Отложенное умение',selector('ready_ability','Умение',choices,null),()=>{
   const a=choices.find(a=>a.id===num(formObject().ready_ability));if(!a)throw Error('Выберите доступное умение');
   setTimeout(()=>abilityForm({...a,use_ready:true,ready_id:b.dataset.ready}),0);
  },'Продолжить');
 }
});

function shiftNotice(c){const shift=c.runtime.initiative_shift;return shift?`<p class="notice">Со следующего раунда: инициатива ${shift.value}${shift.after===c.id?'': ', после '+esc(byId(S.characters,shift.after)?.name||'участника')}.</p>`:''}

document.addEventListener('change',e=>{if(e.target.name==='voluntary_fail'){const input=dialogBody.querySelector('[name=dex_roll]');if(input){input.required=!e.target.checked;input.disabled=e.target.checked}}});
